### DISCUSSION TITLE
[IDEA] Feature flags

### DISCUSSION THREAD

**Discussion by benwebber** [TiddlyWiki/TiddlyWiki5#6484] [Ideas]
Some changes like #6402 are quite self-contained. It would be easy to turn them on or off with a feature flag tiddler, similar to turning on performance instrumentation.

This would be a way to get experimental code in the hands of users without committing to supporting it immediately (and forever). Ideally, most flagged features would make their way into a future release, but that would not be guaranteed.

One could argue that plugins are an equivalent way to deliver experimental code. However, as Cameron pointed out, a filter cache plugin would need to monkey-patch the core. That's less stable than a flagged code path, and harder to integrate over time.

The main downside would be added complexity. We'd also need to pay attention to the cost of looking up the flag value in hot code paths.

Sorry if this has been discussed in the past. I couldn't find a discussion with a cursory search.

**Comment by Jermolene** [comment]
Thanks @benwebber, an interesting idea. I've picked up a little of the excitement about feature flags but haven't given them much thought for TiddlyWiki. Our current approach with novel, risky changes is to implement them in a plugin before committing them to the core (less risky stuff goes into the prerelease directly). But of course there's many core changes that are tricky to implement in that way, some of which have come up quite recently.

Did you have any particular changes in mind for this?

By the way, for the moment, I converted this from an issue to a discussion thread.

**Comment by benwebber** [comment]
At the moment, I'd consider flags for any *changes* to the core JavaScript that would be difficult to encapsulate in a plugin. It's easy enough to add new functions, but it's more difficult to patch existing ones. I would still consider plugins the standard way to test completely new, isolated features like filter operators and macros, or to tweak WikiText macros in the core.

In particular, I'd consider flags for things that have uncertain performance implications. Things like the filter cache, startup modules, or new indexers.

In terms of implementation, I can imagine something like this:

1. Create a namespace under `$:/config` for flags (e.g., `$:/config/FeatureFlags` or `$:/config/Performance` if we want to keep it narrow in scope for now).
2. Add a control panel section to toggle those tiddlers.

    There would be a clear note that these are experimental, and a reminder to back up your wiki.
4. Add a startup module that caches flags toggled by the user in some property of `$tw`.

    Toggling a flag would always require refreshing the wiki. This would have to run quite early, see (4).
5. Within each flagged code path, simply check `if($tw.featureFlags.foo) {}`.

    Modules like indexers are loaded dynamically by scanning the wiki. They could take a new field declaring the flag. If that field is defined and non-empty, we would check if the corresponding flag is enabled before loading the module.

    Or we could always load the indexer, and simply return early in `update()`/`rebuild()`.

**Reply by Jermolene** [reply]
Hi @benwebber

> In terms of implementation, I can imagine something like this:

Sounds exactly right. Perhaps it would be worth exploring a PR that introduced the mechanism you describe, and wires it up to allow the indexers to be enabled/disabled.

**Reply by benwebber** [reply]
Sure, I'll put something together :+1:

**Reply by Jermolene** [reply]
Thanks @benwebber much appreciated
